src/resources/gpio_linux.cc

```diff
@@ -0,0 +1,452 @@
+// Copyright (C) 2024 Toitware ApS.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; version
+// 2.1 only.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// The license can be found in the file `LICENSE` in the top level
+// directory of this repository.
+
+#include "../top.h"
+
+#ifdef TOIT_LINUX
+
+#include <dirent.h>
+#include <errno.h>
+#include <gpiod.h>
+#include <sys/stat.h>
+
+#include "../objects_inline.h"
+#include "../os.h"
+#include "../primitive.h"
+#include "../process.h"
+#include "../resource.h"
+#include "../resource_pool.h"
+#include "../utils.h"
+#include "../vm.h"
+
+#include "../event_sources/epoll_linux.h"
+
+namespace toit {
+
+class GpioResourceGroup : public ResourceGroup {
+ public:
+  TAG(GpioResourceGroup);
+  explicit GpioResourceGroup(Process* process)
+      : ResourceGroup(process){}
+  // TODO(florian): add event handling for the pin's file descriptors.
+};
+
+class GpioChipResource : public Resource {
+ public:
+  TAG(GpioChipResource);
+  GpioChipResource(ResourceGroup* group, gpiod_chip* chip)
+      : Resource(group)
+      , chip_(chip){}
+
+  ~GpioChipResource() override {
+    gpiod_chip_close(chip_);
+  }
+
+  gpiod_chip* chip() { return chip_; }
+
+ private:
+  gpiod_chip* chip_;
+};
+
+class GpioPinResource : public Resource {
+ public:
+  TAG(GpioPinResource);
+  GpioPinResource(ResourceGroup* group, int offset)
+      : Resource(group)
+      , offset_(offset){}
+
+  ~GpioPinResource() override;
+
+  int offset() { return offset_; }
+
+  gpiod_line_settings* settings() { return settings_; }
+  void replace_settings(gpiod_line_settings* settings) {
+    if (settings == settings_) return;
+    if (settings_ != null) gpiod_line_settings_free(settings_);
+    settings_ = settings;
+  }
+
+  gpiod_line_request* request() { return request_; }
+  void set_request(gpiod_line_request* request) { request_ = request; }
+
+  Object* apply_and_store_settings(gpiod_line_settings* settings, Process* process);
+
+ private:
+  int offset_ = -1;
+  gpiod_line_settings* settings_ = null;
+  gpiod_line_request* request_ = null;
+};
+
+static int chip_filter(const struct dirent* entry) {
+  char* path;
+
+  if (asprintf(&path, "/dev/%s", entry->d_name) == -1) return 0;
+
+  int result = false;
```

Comment:
Maybe use 0 and 1 if you're going to return this as an int?

Reply:
done.